### PR TITLE
Alpha Trend Indicator Added to Resolve Issue #658

### PR DESCRIPTION
**Pull request by aligheshlaghi97**
The code in this [link](https://github.com/OnlyFibonacci/AlgoSeyri/blob/main/alphaTrendIndicator.py) works very fine and the outputs are exactly like TradingView's AlphaTrend [here](https://tr.tradingview.com/script/o50NYLAZ-AlphaTrend/). Many thanks to @arifkenan for providing great data about AlphaTrend in issue #658 .
 
I have good news for @arifkenan and @ekremus who are probably waiting for AlphaTrend indicator to be added to pandas-ta, I have implemented and now waiting for @twopirllc to merge.

I have used the idea of the code mentioned above, to implement AlphaTrend taking advantage of Pandas library to accelerate the speed of run and code readability. After all, I have done some benchmark to compare previous code with current one, to see if there is any significant changes, and the results were surprising. 
New implementation have nearly half or fewer time taken to run, here you can see the results, for multiple runs over 10K candlesticks, showing nearly 60ms run time for new implementation vs 140ms for previous code:

Screenshot from 2023-06-04 19-04-02

I also took advantage of Scalene library to find how each of the codes perform based on python, native and system run time. Results were also interesting to pay attention to. Previous code, implemented to use nearly 70% of time inside python code:
Screenshot from 2023-06-04 19-54-36

However, new implementation used nearly 44% of time inside python code, and more time was spent on native code (C language), which means better performance overall:
Screenshot from 2023-06-04 19-53-49

I have also compared outputs of new code with previous code, they were exactly the same over nearly 10K candlesticks for k1 and k2 (which are the two line of AlphaTrend indicator output):
Screenshot from 2023-06-04 19-48-58

### CODE REVIEW

**Comment by arifkenan** [comment]
When will alpha trend indicator implemented to pandas_ta?

**Comment by twopirllc** [comment]
@arifkenan,

> When will alpha trend indicator implemented to pandas_ta?

This pull request has two indicators to be merged, refactored, and tested which is a time consuming process. Thus any expected date or time frame is open ended for any and all bug fixes or feature requests. If the need for the indicator supersedes the unexpected inclusion time, users are encouraged to copy the existing code in the PR and manually insert it into their local source until it is finally incorporated into the library.

> Hi, thank you for all your work. You people are amazing. I was wondering if you can add AlphaTrend Indicator by Kıvanç Özbilgiç?
Because we all getting amazing reviews on this new indicator and maybe you can help us.

Although this library benefits from the contributions of many generous individuals, it continues to face a shortfall in financial support and donations necessary to sustain it as a full-time commitment and cover associated development costs. As a result, I am only able to allocate a portion of my already limited free time to its maintenance and improvement. 😑 

If you and your coworkers are not able to contribute code to enhance and maintain this library **maybe you can help us**. We would like to kindly request your consideration for providing financial assistance through [sponsorship](https://github.com/twopirllc) or at the very least a [donation](https://www.buymeacoffee.com/twopirllc). Your support in this manner would greatly contribute to the continued growth and sustainability of the library. 😎

Kind Regards,
KJ

**Comment by arifkenan** [comment]
still we can't use this indicator?

**Comment by twopirllc** [comment]
@arifkenan,

> still we can't use this indicator?

I have already explained this in the prior comment.

Lastly, as an Open Source Library, you are **not** constrained from using the PR code supplied here and merging it into your code _until_ it has been this has been merged into the _development_ branch.

Your patience and understanding is appreciated! 😎 

Kind Regards,
KJ

File: pandas_ta/trend/alphatrend.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    
+    alpha_ind = DataFrame({'k1': alpha_trend_, 
+                              'k2': alpha_trend_.shift(2).fillna(0)})
+    alpha_ind.index = src.index
+    
+    return alpha_ind
+
+
+alphatrend.__doc__ = \
+""" Alpha Trend Inddicator
```
Comment (ekremus) [inline pandas_ta/trend/alphatrend.py:82]: Indicator 👯

**Review by ekremus — CHANGES_REQUESTED** [review]
""" Alpha Trend Indicator

**Comment by twopirllc** [comment]
@aligheshlaghi97, @arifkenan, @ekremus 

**Alphatrend** is now up on the _development_ branch. If you can @aligheshlaghi97, I would be interested in the performance difference between the _development_ version and the the code you submitted above with Scalene. 🤔 

```sh
$ pip install -U git+https://github.com/twopirllc/pandas-ta.git@development
```

**Zigzag** on the other hand will not be included, as I need to find time to reverse engineer and refactor it.

Kind Regards,
KJ

**Comment by aligheshlaghi97** [comment]
Hi @twopirllc , Thanks for the efforts you put into this library and refactoring my codes. I am actually love to tackle into these problems, and as I checked your code, the time consumed in native C code, has increased to 79.5% (on my PC HP-EliteBook-830-G7-Notebook), which is great enhancement in comparison with my code. You can see the result below:

Screenshot from 2023-10-13 02-19-09


But the main point is on the total time of run, which is nearly 4 times more than the time to run my code (6 ms vs 28 ms). You can see it below:

Screenshot from 2023-10-13 02-44-04

If I could do anything else to help, please don't hesitate to let me know.

**Comment by twopirllc** [comment]
@aligheshlaghi97,

> Thanks for the efforts you put into this library and refactoring my codes.

Thanks! Hope it helps.

I do want to point out from prior indicator speed improvements, that nested functions are more costly to execute because the nested code has to be recompiled each time it is called. That's why I moved the inner functions out. Eventually all remaining indicators with nested functions will be refactored.

> But the main point is on the total time of run, which is nearly 4 times more than the time to run my code (6 ms vs 28 ms). 

Interesting. 🤔

I assume with your test, you have _numba_ installed and ran it _at least once_ **before** running the Scalene test?

Per their [docs](](https://numba.pydata.org/numba-doc/latest/user/5minguide.html)): 
> First, recall that Numba has to compile your function for the argument types given before it executes the machine code version of your function, this takes time. However, once the compilation has taken place Numba caches the machine code version of your function for the particular types of arguments presented. If it is called again the with same types, it can reuse the cached version instead of having to compile again.

> A really common mistake when measuring performance is to not account for the above behaviour and to time code once with a simple timer that includes the time taken to compile your function in the execution time.

KJ

**Comment by aligheshlaghi97** [comment]
@twopirllc,

Thanks for your reply and the points you mentioned.

> I assume with your test, you have _numba_ installed and ran it _at least once_ **before** running the Scalene test?

No I was making a mistake while testing development branch implementation. But I've done this now and tested again, the results show that 71 percent of time spent in native code for your implementation, as you can see below:

Screenshot from 2023-10-13 12-43-03


But the main improvement is the time of run, which is literally 10 times better as you can see below:

Screenshot from 2023-10-13 12-53-26

The running time for your code is 27 ms vs 60 ms for my code. (In previous reply I made a mistake about the time, as I forgot 0 from end of each number):
> > But the main point is on the total time of run, which is nearly 4 times more than the time to run my code (6 ms vs 28 ms).

It was actually 60 ms vs 280 ms but after running some numba code **before** running your implementation, things totally changed and the time to run your code has decreased from 280 ms to only 27 ms. 

Now it seems that the code on development branch is nearly 2 times faster than my implementation. But my question is that how this trick of running numba code before running the main code, has that much of improvement?

**Comment by twopirllc** [comment]
@aligheshlaghi97

> Now it seems that the code on development branch is nearly 2 times faster than my implementation. 

Thanks for double checking for me with Scalene. 😎 Looks like a useful tool. It's good to see the overall speed improvements the library has gotten with _numpy/numba_ being rolled in bit by bit.

> But my question is that how this trick of running numba code before running the main code, has that much of improvement?

I am not a numba expert, but this [A ~5 min guide to Numba](https://numba.readthedocs.io/en/stable/user/5minguide.html) is good.

> How to compile the _numba_ before main?
I think it depends on your environment. For instance in a Jupyter Notebook, you would create a cell to run the All Study on a small test DataFrame. With a enough bars to trigger all the _numba_ indicators. Now if your running in an app in the cloud or a remote server, that's a different situation. Again you would have to compile _numba_ with some small test DataFrame and then you could update an environment variable or create a tmp file and check it's existance or fire off some event message or ... . 🤷🏼‍♂️ 

Luckily _numba_ has a possible solution with [Compiling code ahead of time](https://numba.readthedocs.io/en/stable/user/pycc.html). I haven't had the time to check and test it out, but hopefully this solves that issue. 🤷🏼‍♂️

**Comment by aligheshlaghi97** [comment]
@twopirllc 

Thanks for your message and sharing those links with me. I will study more about numba and hopefully have more effective commits afterwards. Also if I could help with Scalene or profiling tools or methods, do not hesitate to let me know.

Best Regards,
Ali Gheshlaghi

**Pull request closed without merging**